tests/utils/mod.rs

```diff
@@ -0,0 +1,195 @@
+#![allow(unused)] // Different tests use a different subset of functions
+
+use bytes::Bytes;
+use futures::{
+    executor::{block_on, block_on_stream},
+    io::{AsyncBufRead, AsyncRead, AsyncReadExt},
+    stream::{self, Stream},
+};
+use pin_project::unsafe_project;
+use pin_utils::pin_mut;
+use proptest_derive::Arbitrary;
+use std::{
+    io::{self, Cursor, Read},
+    pin::Pin,
+    task::{Context, Poll},
+    vec,
+};
+
+#[derive(Arbitrary, Debug)]
+pub struct InputStream(Vec<Vec<u8>>);
+
+impl InputStream {
+    pub fn stream(&self) -> impl Stream<Item = io::Result<Bytes>> {
+        let stream = stream::iter(self.0.clone().into_iter().map(Bytes::from).map(Ok));
+        PendStream::new(stream)
+    }
+
+    pub fn reader(&self) -> impl AsyncBufRead {
+        // TODO: By using the stream here we ensure that each chunk will require a separate
+        // read/poll_fill_buf call to process to help test reading multiple chunks. This is
+        // blocked on having AsyncBufRead implemented on IntoAsyncRead:
+        // (https://github.com/rust-lang-nursery/futures-rs/pull/1575)
+        //
+        // PendRead::new(self.stream().into_async_read())
+        PendRead::new(Cursor::new(self.bytes()))
+    }
+
+    pub fn bytes(&self) -> Vec<u8> {
+        self.0.iter().flatten().cloned().collect()
+    }
+}
+
+// Happens to be the only dimension we're using
+impl From<[[u8; 3]; 2]> for InputStream {
+    fn from(input: [[u8; 3]; 2]) -> InputStream {
+        InputStream(vec![Vec::from(&input[0][..]), Vec::from(&input[1][..])])
+    }
+}
+
+#[unsafe_project(Unpin)]
+struct PendStream<S>(bool, #[pin] S);
+
+/// Injects at least one Poll::Pending in between each item
+impl<S: Stream> PendStream<S> {
+    fn new(stream: S) -> PendStream<S> {
+        PendStream(false, stream)
+    }
+}
+
+impl<S: Stream> Stream for PendStream<S> {
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        if *this.0 {
+            let next = this.1.poll_next(cx);
+            if next.is_ready() {
+                *this.0 = false;
+            }
+            next
+        } else {
+            cx.waker().wake_by_ref();
+            *this.0 = true;
+            Poll::Pending
+        }
+    }
+}
+
+#[unsafe_project(Unpin)]
+struct PendRead<R>(bool, #[pin] R);
+
+/// Injects at least one Poll::Pending in between each ready read
+impl<R: AsyncBufRead> PendRead<R> {
+    fn new(reader: R) -> PendRead<R> {
+        PendRead(false, reader)
+    }
+}
+
+impl<R: AsyncRead> AsyncRead for PendRead<R> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
```

Comment:
I think we should get a variable that is self explanatory compared for `this`.

Reply:
This should really be `let self = self.project();`, but you can't use `self` as a variable name, so `this` is just a synonym for it.